### PR TITLE
EASY-1802 give write permission to group for deposits 

### PR DESCRIPTION
**Pull request by stephanHogenboom**
Fixes EASY-1802

#### When applied it will
* give write acces to the group
#### How should this be manually tested?

try to modify/ delete a deposit and its content on  deasy/ teasy in the sword tmp folder

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.sword2/DepositHandler.scala [hunk 1]
```diff
@@ -475,36 +475,15 @@ object DepositHandler {
...
 
   def moveBagToStorage(depositDir: File, storageDir: File)(implicit settings: Settings, id: String): Try[File] =
     Try {
       log.debug(s"[$id] Moving bag to permanent storage")
-      if (isOnPosixFileSystem(depositDir))
-        Files.walkFileTree(depositDir.toPath, MakeAllGroupWritable(settings.depositPermissions, id))
-      Files.move(depositDir.toPath.toAbsolutePath, storageDir.toPath.toAbsolutePath).toFile
+      val tempDir = new File(settings.tempDir, id)
+      val storageDir = new File(settings.depositRootDir, id)
+      FilesPermissionService.changePermissionsForDirectoryAndContent(tempDir, settings.depositPermissions, id).get
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.sword2/DepositHandler.scala]: try to write this `Try` block as a for-comprehension. That way you loose the `.get`.

File: src/test/scala/nl.knaw.dans.easy.sword2/FilePermissionServiceSpec.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  private val bagSequenceDir = inputDir / "bag-sequence"
+  private val bagSequenceId = "bag-sequence"
+  private val bagSequenceDirPath = bagSequenceDir.path
+  private val ownerAndGroupRightsString = "rwxrwx---"
+  private val ownerRights = "rxw------"
+
+  override def beforeEach: Unit = {
+    super.beforeEach()
+    inputDir.clear()
+    better.files.File(getClass.getResource("/input/").toURI).copyTo(inputDir)
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.sword2/FilePermissionServiceSpec.scala]: normal import here, please

File: src/test/scala/nl.knaw.dans.easy.sword2/FilePermissionServiceSpec.scala [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    FilesPermissionService.changePermissionsForDirectoryAndContent(bagSequenceDir.toJava, ownerRights, bagSequenceId)
+    Files.getPosixFilePermissions(bagSequenceDirPath) should contain only(
+      PosixFilePermission.OWNER_EXECUTE,
+      PosixFilePermission.OWNER_READ,
+      PosixFilePermission.OWNER_WRITE,
+    )
+  }
+
+
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.sword2/FilePermissionServiceSpec.scala]:
```suggestion
```

**Review by rvanheest — CHANGES_REQUESTED** [review]

**Review by rvanheest — APPROVED** [review]

File: src/main/scala/nl.knaw.dans.easy.sword2/DepositHandler.scala [hunk 4]
```diff
@@ -66,9 +65,10 @@ object DepositHandler {
     val payload = Paths.get(settings.tempDir.toString, id, deposit.getFilename.split("/").last).toFile
     for {
       _ <- assertTempDirHasEnoughDiskspaceMarginForFile(payload)
-      _ <- copyPayloadToFile(deposit, payload)
+      _ <- copyPayloadToFile(deposit, payload) //TODO should file permissions also be set after this action?
       _ <- doesHashMatch(payload, deposit.getMd5)
       _ <- handleDepositAsync(deposit)
+      _ <- FilesPermissionService.changePermissionsForDirectoryAndContent(deposit.getFile, settings.depositPermissions, id) // set file permissions after continued deposit is finished
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.sword2/DepositHandler.scala]: I guess Util is more common in our team than Service. Service reminds too much of servers with servlets.

File: src/main/scala/nl.knaw.dans.easy.sword2/DepositHandler.scala [hunk 5]
```diff
@@ -475,37 +475,14 @@ object DepositHandler {
...
-  def moveBagToStorage(depositDir: File, storageDir: File)(implicit settings: Settings, id: String): Try[File] =
-    Try {
-      log.debug(s"[$id] Moving bag to permanent storage")
-      if (isOnPosixFileSystem(depositDir))
-        Files.walkFileTree(depositDir.toPath, MakeAllGroupWritable(settings.depositPermissions, id))
-      Files.move(depositDir.toPath.toAbsolutePath, storageDir.toPath.toAbsolutePath).toFile
-    }.recover { case e => throw new SwordError("Failed to move dataset to storage", e) }
+  def moveBagToStorage(depositDir: File, storageDir: File)(implicit settings: Settings, id: String): Try[File] = {
+    log.debug(s"[$id] Moving bag to permanent storage")
+    FilesPermissionService.changePermissionsForDirectoryAndContent(depositDir, settings.depositPermissions, id)
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.sword2/DepositHandler.scala]: changePermissionsRecursively?

File: src/main/scala/nl.knaw.dans.easy.sword2/FilesPermissionService.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import scala.util.Try
+
+object FilesPermissionService {
+
+  def changePermissionsForDirectoryAndContent(depositDir: File, permissions: String, id: String): Try[Unit] = Try {
+    if (isOnPosixFileSystem(depositDir))
+      Files.walkFileTree(depositDir.toPath, ChangePermissionsForDirectoryAndContent(permissions, id))
+  }
+
+  // previous classname MakeAllGroupWritable was deceptive as the argument (posix permission string) can be something else than rwxrwx---
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.sword2/FilesPermissionService.scala]: This type of comment belongs in a commit message.

File: src/main/scala/nl.knaw.dans.easy.sword2/FilesPermissionService.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    override def visitFile(path: Path, attrs: BasicFileAttributes): FileVisitResult = {
+      log.debug(s"[$id] Setting the following permissions $permissions on file $path")
+      try {
+        Files.setPosixFilePermissions(path, PosixFilePermissions.fromString(permissions))
+        FileVisitResult.CONTINUE
+      } catch {
+        case usoe: UnsupportedOperationException => log.error("Not on a POSIX supported file system"); FileVisitResult.TERMINATE
+        case cce: ClassCastException => log.error("No file permission elements in set"); FileVisitResult.TERMINATE
+        case ioe: IOException => log.error(s"Could not set file permissions on $path"); FileVisitResult.TERMINATE
+        case se: SecurityException => log.error(s"Not enough privileges to set file permissions on $path"); FileVisitResult.TERMINATE
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.sword2/FilesPermissionService.scala]: For all 4:
either `case _: SecurityException` or 
```suggestion
        case se: SecurityException => log.error(s"Not enough privileges to set file permissions on $path",se); FileVisitResult.TERMINATE
```

**Pull request merged**